updated CDM_LSS to use MUSIC

YuriOku opened the pull request:
In this PR, I have updated the test problem of cosmological N-body simulation, CDM_LSS, to use the MUSIC initial condition generator. 

I have added the output plugin for GAMER in MUSIC, which is available from the main repositories of [MUSIC](https://bitbucket.org/ohahn/music/src/master/) and will soon be also at [MUSIC v2](https://github.com/cosmo-sims/MUSIC2). The CDM_LSS is updated accordingly by adding the configuration file for a CDM N-body simulation. 

I have also added a Python script, `measure_power_spectrum.py`, which computes the matter power spectrum to high-k using [nbodykit](https://nbodykit.readthedocs.io/en/latest/index.html) and a reference matter power spectrum at z=0, `powerspec_z0_Planck18.dat`, which is generated with CAMB. The figure below compares power spectra obtained from the CDM_LSS test problem. The power spectrum obtained by `measure_power_spectrum.py` with 512^3 grid resolution (blue dashed line) shows a nice agreement with the CAMB power spectrum (red dashed line), confirming the initial condition and the cosmological integration.

I have modified `tool/inits/GADGET2_to_PAR_IC.py` to handle multiple particle types in a GADGET file, which happens for a zoom-in initial condition. This is not relevant to the uniform resolution initial condition like the one for CDM_LSS.

![Power_Spectrum_Output](https://github.com/user-attachments/assets/c418b03c-0c8f-459c-99ae-f2e0eb41870c)

hyschive commented:
@YuriOku Nice! Super happy to see this new functionality. The tool for improving the power spectrum calculations at the high-k end is also very helpful.

@koarakawaii Since you are familiar with MUSIC, could you help review and test this PR? Please go through all modifications carefully and also test whether you can reproduce the above figure easily. Once it's done, share the particle initial condition file with me so that I can upload it somewhere for reproducibility. Thanks! I'll also take a close look at this PR after you approve it.

koarakawaii commented:
Hi @YuriOku : I tried using the `ics_music.conf` given in this PR, and fed it to the latest [`MUSIC`](https://bitbucket.org/ohahn/music/src/master/) (commit `6248d133ab20aa84c9cb3b843a8e021a3119be70`). The `PAR_IC` can be generated, but was unable to run that IC with `gamer` (compiled from this PR as well, commit hash `6579ea6da54136646ca5387eebe6107cc67d6161`). The error message is as below:
```
********************************************************************************
ERROR : time-step estimation by particle acceleration is incorrect (dt_acc =           inf, NParAcc = 2097152) !!
        Rank <0>, file <Particle/Par_GetTimeStep_VelAcc.cpp>, line <258>, function <Par_GetTimeStep_VelAcc>
********************************************************************************
```
and I guess it is due to the mass for all the CDM particles inside the `PAR_IC` are `0.0`. But it seems like you can run it smoothly since you can generate the power spectrum, so I am wondering whether it is due to any wrong operation on my side when generating the `PAR_IC`. Can you give me some hints that might result in the zero mass for the CDM particles?

koarakawaii commented:
Below is the input related with particles for `Input__Parameter` when running the `gamer`:
```
# particle (PARTICLE only)
PAR_NPAR                      2097152     # total number of particles (must be set for PAR_INIT==1/3; must be an integer)
PAR_INIT                      3           # initialization option for particles: (1=FUNCTION, 2=RESTART, 3=FILE->"PAR_IC")
PAR_IC_FORMAT                 1           # data format of PAR_IC: (1=[attribute][id], 2=[id][attribute]; row-major) [1]
PAR_IC_FLOAT8                -1           # floating-point precision for PAR_IC (<0: default, 0: single, 1: double) [default: same as FLOAT8_PAR]
PAR_IC_MASS                  -1.0         # mass of all particles for PAR_INIT==3 (<0=off) [-1.0]
PAR_IC_TYPE                   2           # type of all particles for PAR_INIT==3 (<0=off) [-1]
PAR_INTERP                    3           # particle interpolation scheme: (1=NGP, 2=CIC, 3=TSC) [2]
PAR_INTEG                     2           # particle integration scheme: (1=Euler, 2=KDK) [2]
PAR_TR_INTERP                 3           # tracer particle interpolation scheme: (1=NGP, 2=CIC, 3=TSC) [3]
PAR_TR_INTEG                  2           # tracer particle integration scheme: (1=Euler, 2=RK2) [2]
PAR_IMPROVE_ACC               1           # improve force accuracy at patch boundaries [1] ##STORE_POT_GHOST and PAR_INTERP=2/3 ONLY##
PAR_PREDICT_POS               1           # predict particle position during mass assignment [1]
PAR_REMOVE_CELL              -1.0         # remove particles X-root-cells from the boundaries (non-periodic BC only; <0=auto) [-1.0]
OPT__FREEZE_PAR               0           # do not update particles (except for tracers) [0]
PAR_TR_VEL_CORR               0           # correct tracer particle velocities in regions of discontinuous flow [0]
```

koarakawaii commented:
Hi @YuriOku :
* Just wonder whether the error is because I used single precision for the test run, and it seems that in the `MUSIC` source code `src/plugins/output_gamer2.cc`, the line `gamer_unit_mass_     = gamer_unit_density_ * gamer_unit_length_ * gamer_unit_length_ * gamer_unit_length_;` will give value about `2.537689351438944e+44`, which is probably beyond the maximum limit value(`1.7e+38`) for single precision, so `gamer_unit_mass_` becomes `+inf`, and thus caused the error?
* I tried modifying the source code by declaring `gamer_unit_*_` as double instead of `real_t`:
    ```
    double gamer_unit_length_;
    double gamer_unit_time_;
    double gamer_unit_density_;
    double gamer_unit_mass_;
    double gamer_unit_velocity_;
    ```
    and the error seems to be gone. Is this modification has any risk or potential issue? It seems to be fine to me, since `gamer_unit_*` are not really passed into any particle writing function (all these functions take some type `real_t` input variables), so probably won't cause any issue. But just want to confirm with you whether this is correct, and should we do the same modification to `music_unit_*` as well~?

YuriOku commented:
@koarakawaii Thank you for your investigation! It seems I left a bug in the MUSIC plugin, and I agree with your solution. There are no problems with using `double` for those units, and we should use `double` for other internal variables for safety. I will make the modification to MUSIC.

koarakawaii commented:
@YuriOku : thank you for the confirmation as well~!

koarakawaii commented:
@hyschive: the power spectra evolution plot given by the `PAR_IC` produced by the given `ics_music.con
f` is attached. It is not quite straight forward to compare it with the above plot for the initial condition and the most early curve (highest `z`), since `ics_music.conf` generates the initial condition at `z = 24` instead of `z = 20` (if needed, we can modify the `Input__DumpTable` to include another curve for `z = 20`, or just generate `PAR_IC` at that time); meanwhile the remaining curves seem to match quite good. I'll send you the path to `PAR_IC` via `Slack`. The issue I encountered for obtaining this plotting is installing `nbodykit`, which I failed to do so with `python3.12` (seems like it is only compatible up to `python3.6`). Setting up another `python3.6` environment with `minconda` seems to solve the issue.

![Power_Spectrum_Output](https://github.com/user-attachments/assets/939d5ce3-b5d0-46aa-8aeb-d942f75a92f1)

hyschive commented:
@YuriOku Please fix the conflict. Thanks.
@koarakawaii Have you confirmed that this PR works well and the related issue in Grackle has been fixed? If so, please approve this PR.

YuriOku commented:
@hyschive Thanks for reviewing! I have fixed the conflict.

By the way, are you working on Grackle support on the comoving coordinate? Actually, I am already trying it, and it is still in progress. If it is in progress also on your side, we may want to merge our efforts.

hyschive commented:
> By the way, are you working on Grackle support on the comoving coordinate? Actually, I am already trying it, and it is still in progress. If it is in progress also on your side, we may want to merge our efforts.

@YuriOku I'm not working on that. But feel free to let me know if there is anything I may help :)

hyschive commented:
@YuriOku @koarakawaii @ChunYen-Chen We are preparing to merge #348 to support integer particle attributes (e.g., particle type), which will also impact the `PAR_IC` format. Unfortunately, this change is not backward-compatible, as the `PAR_IC` format cannot be determined at runtime.

I suggest merging #348 first, followed by updating both this PR and the MUSIC frontend to align with the new `PAR_IC` format. What do you think?

YuriOku commented:
I agree with your suggestion. Currently, the MUSIC's PAR_IC contain only floating point variables, i.e., particle mass, position, and velocity, but does it affected by the update?

YuriOku commented:
I'm looking at PR https://github.com/gamer-project/gamer/pull/351, but can the IDs be generated in GAMER when loading the files? Also, I would like to see a function like Init_ByFile_User_Ptr for PAR_IC. It is helpful for simulations with star formation, feedback etc from MUSIC's PAR_IC.

hyschive commented:
> Currently, the MUSIC's PAR_IC contain only floating point variables, i.e., particle mass, position, and velocity, but does it affected by the update?

If that's the case, there’s no need to update `PAR_IC`. However, I’m confused because `PAR_IC_TYPE` is still set to `-1` in `Input__Parameter`. Shouldn’t it be updated to `2` to represent `PTYPE_DARK_MATTER`?

> but can the IDs be generated in GAMER when loading the files?

Are you referring to particle "types" or "IDs"? If all particles share the same type, you can set it using the runtime option `PAR_IC_TYPE`, as mentioned above, and exclude it from `PAR_IC`. Regarding particle IDs, #351 will support assigning them automatically, and I plan to merge it after #348.

> I would like to see a function like Init_ByFile_User_Ptr for PAR_IC. It is helpful for simulations with star formation, feedback etc from MUSIC's PAR_IC.

Could you clarify why `PAR_IC` alone isn’t sufficient? If it’s for supporting multiple particle types, these can be stored directly in `PAR_IC`.

@koarakawaii @ChunYen-Chen

YuriOku commented:
> I’m confused because PAR_IC_TYPE is still set to -1 in Input__Parameter. Shouldn’t it be updated to 2 to represent PTYPE_DARK_MATTER?

That is a typo. As [commented](https://github.com/gamer-project/gamer/pull/338#issuecomment-2249350287) by @koarakawaii, PAR_IC_TYPE must be set to 2. I'll fix this soon.

> Are you referring to particle "types" or "IDs"? If all particles share the same type, you can set it using the runtime option PAR_IC_TYPE, as mentioned above, and exclude it from PAR_IC. Regarding particle IDs, https://github.com/gamer-project/gamer/pull/351 will support assigning them automatically, and I plan to merge it after https://github.com/gamer-project/gamer/pull/348.

Understood. I was referring to IDs. 

> Could you clarify why PAR_IC alone isn’t sufficient? If it’s for supporting multiple particle types, these can be stored directly in PAR_IC.

In my case, I use the MUSIC IC to simulate star formation and feedback. In my custom GAMER code, there are particle attributes for this subgrid physics, e.g., particle creation time and feedback flags. GAMER expects these attributes to be stored in PAR_IC, but MUSIC's PAR_IC does not have these attributes. I actually set NParAtt=7 by hand in Par_Init_By_File.cpp, but I thought there might be a better way to handle PAR_IC when the number of particle attributes is changed by code extension. Just adding a parameter like PAR_IC_NVAR may be enough for my case.

YuriOku committed:
Update Input__Parameter

koarakawaii commented:
Hi @YuriOku : since the `PAR_IC_TYPE` is set to `2` in the `Input__Paramter`, the old `PAR_IC` downloaded from the `download_ic.sh` might not be applicable anymore (there are `8` attributes in the original file, while we only need `7` now). We have modified that file and uploaded to `ytHub`, which can be downloaded via:
```
curl https://girder.hub.yt/api/v1/item/67767f85999605c485c8de43/download -o PAR_IC 
```
if you don't mind, maybe we can use this new URL to substitute the old URL in `download_ic.sh`? We have tested the output of old `PAR_IC` and modified `PAR_IC` by plotting the `x` project plot, and they look nearly the same (they were run with different MPI ranks without enabling `BITWISE_REPRODUCIBILITY`).
* Old `PAR_IC`
![Data_000074_Particle_x_particle_mass](https://github.com/user-attachments/assets/adfb452b-9664-41b0-9ac1-706dcba752dd)

* New `PAR_IC`
![Data_000074_Particle_x_particle_mass](https://github.com/user-attachments/assets/0fa286e9-df7a-4cc4-92c2-f4bee7c7efe9)

YuriOku committed:
Update download_ic.sh

YuriOku commented:
Hi @koarakawaii, thank you. I have updated the download_ic.sh

koarakawaii commented:
@YuriOku : thank you for the prompt update~

hyschive commented:
@YuriOku Please resolve the conflict.

hyschive commented:
> In my case, I use the MUSIC IC to simulate star formation and feedback. In my custom GAMER code, there are particle attributes for this subgrid physics, e.g., particle creation time and feedback flags. GAMER expects these attributes to be stored in PAR_IC, but MUSIC's PAR_IC does not have these attributes. I actually set NParAtt=7 by hand in Par_Init_By_File.cpp, but I thought there might be a better way to handle PAR_IC when the number of particle attributes is changed by code extension. Just adding a parameter like PAR_IC_NVAR may be enough for my case.

Thanks for the elaboration. I prefer `Par_Init_ByFile_User_Ptr` over `PAR_IC_NVAR`, as the latter doesn’t allow for specifying which attributes to skip. Here is the feature request: #397.

@ChunYen-Chen Please assist with the implementation. It should be very straightforward.

YuriOku committed:
Merge branch 'main' into MUSIC_CDM_LSS

YuriOku commented:
@hyschive resolved the conflict

koarakawaii reviewed (commented):
Hi @YuriOku:
Sorry for the late reply, and thank you for the contribution. The PR looks nice to me, just with some minor comments left as attached. I think the major changes we would like to adopt for the `CDM_LSS` test problem of this PR are:
* Attach a `PAR_IC` with cosmology parameters consistent with the `Input__Paramter` you used. This file has been uploaded to `https://girder.hub.yt` and can be downloaded via the URL in the comment left in `download_ic.sh`
* Update the URL link for the `GADGET2` `hdf5` data for testing the `GADGET2_to_PAR_IC.py`. It seems that I cannot download the old data, so I updated a new `Data_000.hdf5` which is produced by `GADGET2` (with initial condition generated by `MUSIC` with the same cosmology parameters in this PR). Thus, we might need to provide the URL for downloading it in `GADGET2_to_PAR_IC.py` (see comment therein)
* Provide the `input_powerspec.txt` generated by `MUSIC`, such that `plot_script/power_specrum/plot_power_spectrum.py` can work. I attached the file thus generated by `MUSIC` (whose commit is `6248d133ab20aa84c9cb3b843a8e021a3119be70`), hoping that it might save you some time without running `MUSIC` again
*  [Optional] I am not sure whether the `powerspec_z0_Planck.dat` is the post-processed version (by dividing $8\pi^3$ for normalization) of the `*_matterpower.dat` generated by `CAMB`. If this is the case, I think it will be nice that we either provide the post-processing script for normalization, or we can just provide the `*_matterpower.dat` generated by `CAMB` with `plot_script/power_specrum/plot_power_spectrum.py` handling the normalization inline. I am just thinking that if users wants to apply their own cosmology parameters different from here, the plot script can still work, since they should be able to generate their own `*_matterpower.dat` via `CAMB`. I attached the `*_matterpower.dat` generated by my `CAMB` with cosmology parameters set the same as this test problem (renamed as `planck_2018_CDM_LSS_matterpower.txt` and attached). But I guess the safest way is using the `CAMB` matter power spectrum generated from your side~

Hope these modifications will not bother you too much~

[input_powerspec.txt](https://github.com/user-attachments/files/18346463/input_powerspec.txt)
[planck_2018_CDM_LSS_matterpower.txt](https://github.com/user-attachments/files/18346465/planck_2018_CDM_LSS_matterpower.txt)

hyschive commented:
> Thanks for the elaboration. I prefer `Par_Init_ByFile_User_Ptr` over `PAR_IC_NVAR`, as the latter doesn’t allow for specifying which attributes to skip. Here is the feature request: #397.

@YuriOku We have merged #405 to support `Par_Init_ByFile_User_Ptr`.

YuriOku committed:
GADGET2_to_PAR_IC.py

YuriOku committed:
reflect review comments

YuriOku committed:
align =

YuriOku committed:
align =

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+use_LLA			= no
+periodic_TF		= yes
+blocking_factor = 8
+
```
aligned =

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+nspec			= 0.9665
+transfer		= eisenstein
+
+#ZeroRadiation = true # Enable this when using version 2 of MUSIC
```
aligned `=`

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/plot_power_spectrum.py:
```diff
@@ -8,20 +8,28 @@
 k_music          = data_music[:,0]
 PS_music         = data_music[:,-2]
 
+data_camb        = np.loadtxt("powerspec_z0_Planck18.dat")
+k_camb           = data_camb[:,0]
+PS_camb          = data_camb[:,1]
```
I have added a python script `run_camb.py`, which generates `powerspec_z0_Planck18.dat`

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+                partype = np.append(partype, partype_i)
 
                 print('Number of particles = %i'%NumPart_i)
 
```
`GADGET2_to_PAR_IC.py` now prints recommended GAMER parameters. I believe this functionality helps users choosing parameter values, including `PAR_IC_FORMAT`.

YuriOku committed:
Update README

YuriOku committed:
Update README

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+boxlength		= 30
+zstart			= 24   # choosing a somewhat low redshift to evaluate the initial condition
+levelmin		= 7
+levelmin_TF = 7
```
aligned `=`

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+[output]
+format      = gamer2
+filename    = UM_IC   # not used for CDM_LSS test problem
+parfilename = PAR_IC
```
aligned `=`

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/plot_power_spectrum.py:
```diff
@@ -8,20 +8,28 @@
 a_dump           = a_dump[:,1]
 
-data_music       = np.loadtxt("/work1/koarakawaii/music/bin/GENERATE_GADGET_for_CDM/box_length=30.0/input_powerspec.txt")
+data_music       = np.loadtxt("../../input_powerspec.txt")
```
added `input_powerspec.txt` generated by MUSIC in `plot_script/power_spectrum` and modified the line 14 accordingly

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/plot_power_spectrum.py:
```diff
@@ -31,6 +39,8 @@
 plt.subplots_adjust(wspace=0.25)
 ax1 = plt.subplot(121)
 ax1.loglog(k_music, PS_music, c='k', ls="--", label="MUSIC Input PS", lw=0.5)
+ax1.loglog(k_camb, PS_camb, c='r', ls="--", label="CAMB PS (z=0; Planck18)", lw=0.5)
```
updated the code following your comments

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -89,6 +89,14 @@
+        velx = np.array([], dtype=output_float_precision)
+        vely = np.array([], dtype=output_float_precision)
+        velz = np.array([], dtype=output_float_precision)
+        partype = np.array([], dtype=output_float_precision)
```
- added an option to disable outputting the particle types
- updated `input_GADGET_file`

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+
+                # For GADGET, v_peculiar = sqrt(a) * v_snapshot 
+                # --> See "3. Particle velocities" on p.32 of https://wwwmpa.mpa-garching.mpg.de/gadget/users-guide.pdf
+                # For GAMER, v_code = a * v_physical
```
changed v_physical to v_peculiar

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+                posy_i = PartType_Coordinates[:,1]*pos_conversion
+                posz_i = PartType_Coordinates[:,2]*pos_conversion
+
+                # For GADGET, v_peculiar = sqrt(a) * v_snapshot 
```
removed the trailing space

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
                 # For GAMER, we fix the imported particle type to "1 (PTYPE_GENERIC_MASSIVE)."
                 # --> See https://github.com/gamer-project/gamer/wiki/Initial-Conditions#IC-File-Particles
-                partype = np.full(NumPart_i, 1)
+                partype_i = np.full(NumPart_i, 1)
```
I have added an option `output_particle_type`. If it is "True", the particle types are copied from the GADGET input to PAR_IC.

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+        output.write(velx.astype(dtype=output_float_precision).tobytes())
+        output.write(vely.astype(dtype=output_float_precision).tobytes())
+        output.write(velz.astype(dtype=output_float_precision).tobytes())
+        output.write(partype.astype(dtype=output_float_precision).tobytes())
```
added the option

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+                velz_i = Velocities_Coordinates[:,2]*vel_conversion*(input_a_scale_factor**1.5)
 
                 # For GADGET, see "Table 3" on p.31 of https://wwwmpa.mpa-garching.mpg.de/gadget/users-guide.pdf
                 # For GAMER, we fix the imported particle type to "1 (PTYPE_GENERIC_MASSIVE)."
```
now the particle types are copied from GADGET input to PAR_IC if `output_particle_type` is "True"

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+# configuration file of MUSIC cosmological initial condition generator
+# MUSIC can be downloaded from https://bitbucket.org/ohahn/music/src/master/ (version 1) 
```
removed the trailing space

YuriOku commented on example/test_problem/Hydro/CDM_LSS/ics_music.conf:
```diff
@@ -0,0 +1,57 @@
+levelmin		= 7
+levelmin_TF = 7
+levelmax		= 7
+padding			= 8 
```
removed the trailing space

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
 
-                mass = np.full(NumPart_i, GADGET_MassTable[i])*mass_conversion
+                if (GADGET_MassTable[i] == 0.0):
+                    mass_i = np.array(f[key_NOW]["Masses"], float)*mass_conversion
```
changed following your suggestion

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/measure_power_spectrum.py:
```diff
@@ -0,0 +1,48 @@
+particle_catalog = RandomCatalog(csize=npar) # overwritten by the actual values below
+
+particle_catalog['Mass'] = mass
+particle_catalog['Position'] = pos
```
added the link as a comment

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/measure_power_spectrum.py:
```diff
@@ -0,0 +1,48 @@
+particle_catalog['Position'] = pos
+
+# create a mesh object
+mesh = particle_catalog.to_mesh(Nmesh=Nmesh, BoxSize=boxsize)#, value='Mass')
```
added the link as a comment

YuriOku commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/measure_power_spectrum.py:
```diff
@@ -0,0 +1,48 @@
+
+# compute the power spectrum
+boxlen = boxsize[0]
+r = FFTPower(mesh, mode='1d', kmin=1/boxlen, kmax=Nmesh/boxlen, dk=0.1)
```
- changed the parameters of `FFTPower` following your suggestions
- added the links

YuriOku commented on example/test_problem/Hydro/CDM_LSS/Input__Parameter:
```diff
@@ -61,9 +61,9 @@ PAR_TR_VEL_CORR               0           # correct tracer particle velocities i
-HUBBLE0                       0.6732117               # dimensionless Hubble parameter (currently only for converting ELBDM_MASS to code units)
+A_INIT                        0.04  # initial scale factor
+OMEGA_M0                      0.3111                # omega matter at the present time (consistent with GADGET-2 input hdf5 file)
+HUBBLE0                       0.6766               # dimensionless Hubble parameter (currently only for converting ELBDM_MASS to code units)
```
aligned the comments

YuriOku commented on example/test_problem/Hydro/CDM_LSS/download_ic.sh:
```diff
@@ -1 +1 @@
-curl -LJO https://use.yt/upload/c2b0f76b
+curl https://girder.hub.yt/api/v1/item/67767f85999605c485c8de43/download -o PAR_IC 
```
updated the download link

YuriOku commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -97,33 +105,49 @@
+                velx = np.append(velx, velx_i)
+                vely = np.append(vely, vely_i)
+                velz = np.append(velz, velz_i)
+                partype = np.append(partype, partype_i)
```
added a flag option `output_particle_type`

YuriOku commented:
@hyschive @ChunYen-Chen Thanks for supporting `Par_Init_ByFile_User_Ptr`!

YuriOku commented on example/test_problem/Hydro/CDM_LSS/README:
```diff
@@ -26,7 +26,7 @@ Note:
 1. CDM cosmological large-scale structure simulations
 
-2. Fiducial PAR_IC file can be downloaded with the command: sh download_ic.sh
+2. Fiducial PAR_IC file can be generated using MUSIC (https://bitbucket.org/ohahn/music/src/master/) with the parameter file ics_music.conf
```
I restored the previous line. Now, the first option is to download the `PAR_IC`, and users optionally can generate an initial condition by themselves using `ics_music.conf`. 
I am uncertain about how much of an introduction we should provide to MUSIC and its parameters...

YuriOku commented:
Hi @koarakawaii, thank you for your review and sorry for my late reply. 

Thank you for preparing the files. I have updated the links in download scripts. I have also addressed your points. I have added a Python script `run_camb.py`, which runs CAMB and apply the 1/(8 pi^3) factor, in `plot_script/power_spectrum`. I  believe this is helpful if users want to try other cosmological parameters.

hyschive commented:
@YuriOku Please resolve the conflict and update to the latest `main` branch.

@koarakawaii Could you take another look at this PR? If everything looks fine, please approve it. Thanks.

YuriOku committed:
Merge remote-tracking branch 'upstream/main' into MUSIC_CDM_LSS

YuriOku committed:
fix download_ic.sh

YuriOku commented:
@hyschive @koarakawaii, I have resolved the conflict and updated the latest main branch.

hyschive commented:
@YuriOku Apologies for the extremely delayed PR review! I plan to resume this work and complete the review asap.

Could you update this PR to resolve the conflicts again? Thanks.

@koarakawaii If everything looks fine to you, please go ahead and approve this PR. Thanks!

YuriOku committed:
Merge branch 'main' into MUSIC_CDM_LSS

YuriOku commented:
@hyschive Thank you for your review!

@koarakawaii I have resolved the conflict.

HsunYeong commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/measure_power_spectrum.py:
```diff
@@ -0,0 +1,51 @@
+particle_catalog['Mass'] = mass
+particle_catalog['Position'] = pos
+
+# create a mesh object 
```
```suggestion
# create a mesh object
```

HsunYeong commented on example/test_problem/Hydro/CDM_LSS/plot_script/power_spectrum/run_camb.py:
```diff
@@ -0,0 +1,64 @@
+    elif cosmology == "Planck18":
+      self.H0 = 67.66
+      self.ombh2 = 0.02242
+      self.omch2 = 0.11933  
```
```suggestion
      self.omch2 = 0.11933
```

HsunYeong commented on example/test_problem/Hydro/CDM_LSS/Input__Parameter:
```diff
@@ -49,7 +49,7 @@ PAR_IC_FORMAT                 1           # data format of PAR_IC: (1=[attribute
 PAR_IC_FLOAT8                 0           # floating-point precision for PAR_IC (<0: default, 0: single, 1: double) [default: same as FLOAT8_PAR]
```
```suggestion
PAR_IC_FLOAT8                 1           # floating-point precision for PAR_IC (<0: default, 0: single, 1: double) [default: same as FLOAT8_PAR]
```

HsunYeong commented on example/test_problem/Hydro/CDM_LSS/README:
```diff
@@ -19,14 +19,15 @@ Default setup:
 1. CDM cosmological large-scale structure simulations
 
 2. Fiducial PAR_IC file can be downloaded with the command: sh download_ic.sh
+   --> Optionally, PAR_IC file can be generated using MUSIC (https://bitbucket.org/ohahn/music/src/master/) with the parameter file ics_music.conf
```
Maybe we should also ask the user to compile MUSIC in double precision mode to prevent the overflow problem.

HsunYeong commented on example/test_problem/Hydro/CDM_LSS/download_ic.sh:
```diff
@@ -1,11 +1,12 @@
+FILE_SHA256="80549dc26c6896b0f1efb76a378aec05d5fb983b22d6fd0c5323862e5ed2b7b9"
 
 # 1. download
 curl -L https://hub.yt/api/v1/item/${FILE_ID}/download -o ${LOCAL_FILENAME}
```
```suggestion
FILE_ID="69e9d7524a11f772ce54c597"
FILE_SHA256="1c06000f3ae1d889434ec4d2f28fcb502dd73493eb28e0ebca2e54d072128afb"

# 1. download
curl -L https://girder.hub.yt/api/v1/file/${FILE_ID}/download -o ${LOCAL_FILENAME}
```
We updated the link to download the double precision `PAR_IC`.

HsunYeong reviewed (commented):
@YuriOku Sorry for the belated review. 

Since the music overflow fix is still not merged yet, we conclude that we should ask the user to run music in double precision setup to prevent this problem. Therefore we decide to make the default `PAR_IC` double precision and related links and options should be updated. The rest of the PR looks good to me.

YuriOku committed:
Apply suggestions from code review

Co-authored-by: Hsun-Yeong Yang <50362009+HsunYeong@users.noreply.github.com>

YuriOku committed:
Update README with MUSIC compilation instructions

Added note about compiling MUSIC in double precision mode to avoid overflow.

YuriOku commented:
Hi @HsunYeong, thank you for your review. I agree with your decision and have reflected your suggestions in the code.

hyschive commented:
@YuriOku Thanks for the swift response. Could you also update this PR to synchronize with the latest `main` branch? Thanks!

 I'll take one final look at this PR before merging.

YuriOku committed:
Merge remote-tracking branch 'upstream/main' into MUSIC_CDM_LSS

YuriOku commented:
@hyschive I have merged the `main` branch to this branch!

hyschive commented:
> @hyschive I have merged the `main` branch to this branch!

@YuriOku It still appears to be out of sync with `main` somehow. I noticed your latest commit, `Merge remote-tracking branch 'upstream/main' into MUSIC_CDM_LSS`, but it does not seem to have introduced any changes. Could you please double-check? Thanks!

hyschive commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -5,8 +5,10 @@
-                                         # --> an example "Data_000.hdf5" (128^3 particles) can be downloaded via "curl -JO http://use.yt/upload/1d18ed50"
+                                         # --> an example "Data_000.hdf5" (128^3 particles) can be downloaded via "curl https://girder.hub.yt/api/v1/file/677e5577999605c485c8de96/download -o Data_000.hdf5"
+                                         # (MUSIC commit 6248d133ab20aa84c9cb3b843a8e021a3119be70)
 output_float_precision = np.float32      # "np.float32" or "np.float64"
```
How about switching to `np.float64` to match the default?

hyschive commented on tool/inits/GADGET2_to_PAR_IC.py:
```diff
@@ -89,41 +91,83 @@
 
-    print('PAR_IC complete')
+print('PAR_IC complete\n')
+print('Recommended GAMER Input__Parameter settings:')
```
How about also outputting `BOX_SIZE` from `Header_list.attrs["BoxSize"]`? Remember to account for the length-unit conversion.

hyschive commented on tool/inits/GADGET2_to_PAR_IC.py:
I encountered the following error when testing the generated `PAR_IC` (converted from `curl https://girder.hub.yt/api/v1/file/677e5577999605c485c8de96/download -o Data_000.hdf5`) with GAMER:
```
Par_Aux_InitCheck ...
********************************************************************************
ERROR : Pos[0][132212] =  3.0000002e+01 lies outside the simulation domain (0.0 ... 3.0000000e+01) !!
        Rank <1>, file <Particle/Par_Aux_InitCheck.cpp>, line <67>, function <Par_Aux_InitCheck>
********************************************************************************
```

Could this be caused by rounding errors? Note that it occurs for both `output_float_precision = np.float32` and `np.float64`.

hyschive commented on example/test_problem/Hydro/CDM_LSS/README:
Consider briefly mentioning compatibility with `MUSIC2`. Besides, do we need to provide another `ics_music.conf` file for `MUSIC2`?

YuriOku committed:
Merge remote-tracking branch 'upstream/main' into MUSIC_CDM_LSS

YuriOku committed:
Fix alignment

YuriOku commented:
@hyschive I've updated to the latest version. Hope it works now...

hyschive commented on example/test_problem/Hydro/CDM_LSS/download_ic.sh:
I noticed the following warning when testing this initial condition with GAMER. Could it be caused by unit conversion or by small differences in some physical constants defined in `include/PhysicalConstant.h`?

The same warning appears if I use the re-generated `PAR_IC` file from `MUSIC`.

After confirming the cause, either fix it or at least add a brief comment in the `README` about this minor inconsistency.

```
Poi_GetAverageDensity ...
WARNING : background density deviates from unity by 9.04462147143237e-05 (UNITY is assumed in COMOVING) !!
NOTE : background density = 9.99909553785286e-01
Poi_GetAverageDensity ... done
```

hyschive commented on example/test_problem/Hydro/CDM_LSS/README:
Briefly describe all files and scripts in `plot_script/power_spectrum`.

hyschive reviewed (commented):
@YuriOku Thanks for the contribution! I've added some minor comments and will merge this PR once they are addressed.
